### PR TITLE
WIP: Integrate demos

### PR DESCRIPTION
**Pull request by cgcostume**

### CODE REVIEW

File: demos/demos.json [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        "authors": [
+            "anne-gropler",
+            "cgcostume"
+        ],
+        "caption": "High-performance, text placement, typesetting, and rendering of in 2D and 3D.",
+        "disabled": true
+    },
+    {
+        "target": "webvr-support",
+        "title": "WebVR Support",
```
Comment (bwasty) [inline demos/demos.json]: I'd use `WebXR` everywhere - `WebVR` is legacy and will only be supported via [polyfill](https://github.com/immersive-web/webxr-polyfill).

File: package.json [hunk 2]
```diff
@@ -68,19 +68,23 @@
...
         "lint": "tslint --project tsconfig.json --format stylish --config tslint.json",
         "test": "mocha",
         "test-nyc": "nyc mocha",
         "test-watch": "mocha --watch",
         "coverage": "nyc report --reporter=text-lcov | coveralls",
         "doc": "typedoc source/webgl-operate.ts",
-        "ci": "run-p build-dist build-lib lint test-nyc",
+        "pug": "node pugconfig.js",
+        "pug-watch": "node pugconfig.js --watch",
+        "ci": "run-p build-dist build-lib lint test-nyc website",
```
Comment (bwasty) [inline package.json]: I'm not sure about running `test-nyc` and `website` indirectly when npm-installing from git (via `prepack` and `preversion`). Takes quite some extra time though only `build-dist` and `build-lib` are required in that case.

File: source/text/position2dlabel.ts [hunk 3]
```diff
@@ -139,6 +139,7 @@ export class Position2DLabel extends Label {
         }
         vec2.max(this._frameSize, size, [1.0, 1.0]);
         this._altered.alter(this._type);
+        console.log(this._altered.dynamic, this._altered.static);
```
Comment (bwasty) [inline source/text/position2dlabel.ts]: remove?

File: source/text/typesetter.ts [hunk 4]
```diff
@@ -47,11 +47,11 @@ export class Typesetter {
...
         As a consequence do strictly rely on in-place operations only. */
 
-        const padding = fontFace.glyphTexturePadding;
+        // const padding = fontFace.glyphTexturePadding;
         const origin: vec3 = vertices.origin(index);
         vec3.set(origin, pen[0], pen[1], 0.0);
-        origin[0] += glyph.bearing[0] + padding[3];
-        origin[1] += glyph.bearing[1] - glyph.extent[1] + padding[2];
+        origin[0] += glyph.bearing[0]; // + padding[3];
+        origin[1] += glyph.bearing[1] - glyph.extent[1]; // + padding[2];
```
Comment (bwasty) [inline source/text/typesetter.ts]: remove outcommented code? Also below (line 392).

File: website/demos/cornell-box.pug [hunk 5]
```diff
@@ -0,0 +1,16 @@
+extends demo.pug
+
+
+block canvas
+
+  canvas(id = 'demo-canvas', data-backend = 'auto').img-fluid.w-100
+
+
+block breakdown
+  p This should describe the demo in much more detail, e.g., referring to relevant code snippets and resources
```
Comment (bwasty) [inline website/demos/cornell-box.pug]: No placeholders please - actual description or leave it out for now (same for cubescape below).

File: website/partials/demo-thumbnail.pug [hunk 6]
```diff
@@ -0,0 +1,18 @@
+include authors.pug
+
+mixin demo-thumbnail(source, title, authors, caption, disabled)
+  div(class = disabled ? 'disabled text-muted' : '').card
+    if disabled
+      img(src = 'img/' + source + '.jpg').img-fluid
+      h4.badge-parent #[span.badge.badge-dark in progress]
+    else
+      a(href = '/demos/' + source + '.html').card-img-top #[img(src = 'img/' + source + '.jpg').img-fluid]
```
Comment (bwasty) [inline website/partials/demo-thumbnail.pug]: Do we really need the PNG versions of all images when only the JPG is used?

File: website/index.pug [hunk 7]
```diff
@@ -1,37 +1,65 @@
...
+    .row
+      .col-lg-6.col-12.pb-5
+        img(src = 'img/webgl-operate-logo.svg' alt = 'webgl-operate logo').mb-2.w-100.d-none.d-md-block
+        img(src = 'img/webgl-operate-logo.svg' alt = 'webgl-operate logo').mb-2.d-block.d-md-none
+
+        p.lead Build high-performance rendering applications with this use case agnostic, TypeScript based WebGL engine.
+        p.lead.mb-4 webgl-operate is an open source engine that enables hardware-accelerated 2D and 3D rendering and computation. It is highly extensible and facilitates arbitrary rendering approaches such as forward rendering, offscreen computing, multi-frame sampling, raymarching raytracing, or progressive pathtracing.
+
+        a(href = '/doc/' role = 'button').btn.btn-lg.btn-primary.ml-4.mr-2 Get Started
+        a(href = 'https://github.com/cginternals/webgl-operate/archive/v0.4.2-beta.zip' role = 'button').btn.btn-lg.btn-primary.ml-2 Download
```
Comment (bwasty) [inline website/index.pug]: `v0.4.2-beta.zip`?

File: website/partials/navbar.pug [hunk 8]
```diff
@@ -1,18 +1,32 @@
+- var navLinks = []
+- navLinks.push({ href: '/', text: 'webgl-operate' })
+- navLinks.push({ href: '/examples/', text: 'Examples', disabled: true })
```
Comment (bwasty) [inline website/partials/navbar.pug]: Just disabling it looks a bit weird - perhaps add `coming soon`?

File: website/index.pug [hunk 9]
```diff
@@ -1,37 +1,65 @@
...
+            +demo-thumbnail(demo.target, demo.title, demo.authors, demo.caption, demo.disabled)
+
 
   section.container
     div.row
       div.col
+        h2 Your WebGL Status
+
         include partials/code.pug
         +code('context-about', ' ', 'context.about:')
```
Comment (bwasty) [inline website/index.pug]: I'm not sure it's a good idea to integrate this on the main landing page.

File: website/index.pug [hunk 10]
```diff
@@ -1,37 +1,65 @@
...
+        +code('jsdelivr', '<script src="https://cdn.jsdelivr.net/npm/webgl-operate@latest/dist/webgl-operate.min.js"></script>'
+          , 'You can integrate webgl-operate as bundled JavaScript in your HTML:')
+        +code('npm', 'npm install webgl-operate'
+          , 'Or you can use webgl-operate as JavaScript or TypeScript library (types included) via npm:', 'language-sh')
+
+
+  section.container#demos
+    div.row
+      div.col
+        h2 Demos
```
Comment (bwasty) [inline website/index.pug]: Suggestion: link to https://github.com/cginternals/webgl-operate/tree/integrate-demos/demos here, or to the individual demo source on the subpages.

File: website/index.pug (hunk 9 above)
Comment (cgcostume) [reply website/index.pug]: not sure either, but i have no idea how to link the about page... per demo? only link to plain about? ...

File: website/partials/navbar.pug (hunk 8 above)
Comment (cgcostume) [reply website/partials/navbar.pug]: coming very soon ... integrated! 😄

File: website/index.pug (hunk 7 above)
Comment (cgcostume) [reply website/index.pug]: hmm.. some npm packaging mess, i tried to use postfixes/tags... should be resolved with next release

File: website/partials/demo-thumbnail.pug (hunk 6 above)
Comment (cgcostume) [reply website/partials/demo-thumbnail.pug]: hmm. where to store them? i like to have the originals nearby.

File: source/text/typesetter.ts (hunk 4 above)
Comment (cgcostume) [reply source/text/typesetter.ts]: this is some leftovers from another branch... will be removed in line anchor fix.

**Pull request merged**